Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/dialogController/UserMenuController.java

```diff
@@ -213,17 +213,19 @@ public void onEvent(Event event) {
                         try {
                             boolean logoutSuccess = false;
 
+                            final Session session = Sessions.getCurrent();
+
                             if (config.isUseKeycloakSso()) {
-                                logoutSuccess =
-                                        managerService.logoutUserAllSessions(
-                                                currentUser.getUsername(),
-                                                config.getSecurityMsHttpLogoutUrl(),
-                                                config.getSecurityMsHttpsLogoutUrl());
-                                LOGGER.info("\nlogoutSuccess: {}", logoutSuccess);
+                                if (event.getData().equals(session)) {
+                                    logoutSuccess =
+                                            managerService.logoutUserAllSessions(
+                                                    currentUser.getUsername(),
+                                                    config.getSecurityMsHttpLogoutUrl(),
+                                                    config.getSecurityMsHttpsLogoutUrl());
+                                    LOGGER.info("\nlogoutSuccess: {}", logoutSuccess);
+                                }
                             }
 
-                            final Session session = Sessions.getCurrent();
-
                             if ((session == null || event.getData().equals(session)) || (logoutSuccess)) {
                                 Clients.evalJavaScript("window.close()");
                                 Executions.sendRedirect("/j_spring_security_logout");
```

Comment:
Should the processing on ZK window and redirect to Spring only be executed if "logoutSuccess = true"?

What happens if the managerService.logoutUserAllSessions fails for some reason (REST endpoint fails) and logoutSuccess = false, so the current user doesn't log out Keycloak but ZK window is closed and Spring logout is executed?

Reply:
It's not recommended to put the logout logic inside UI event listener. It'd better put in a separate logic class to manage the logout processing logic, e.g. process keycloak logout, process ZK session, redirect to Spring logout, etc. It helps build an abstraction layer on top of these various misc. details; meanwhile also support unit tests.

Reply:
Why doing both Clients.evalJavaScript("window.close()") and Executions.sendRedirect("/j_spring_security_logout")? The former will close the window (browser tab) and so it won't be able to show the login page in the same window.

In addition, putting redirect in this kind of event listener will lead to too many redirect errors. Imagine opening many browser tabs, each can be multiple Dashboards, multiple PDs, log animation, and so when they respond to a publish "sign out" event, there are many redirect requests.

![image](https://user-images.githubusercontent.com/12888841/119864053-95249b00-bf5d-11eb-8c93-d0c2be19a714.png)

Reply:
> Why doing both Clients.evalJavaScript("window.close()") and Executions.sendRedirect("/j_spring_security_logout")? The former will close the window (browser tab) and so it won't be able to show the login page in the same window.

It's tricky and inobvious.  The time this code was written, browsers would only allow "window.close()" to succeed for tabs that had been opened by scripted action such the BPMN editor, PD, etc.  You couldn't close a window the user had directly opened like the Portal.  By doing it this way, all the tabs except for the Portal would be closed, and the Portal would show the login page.  Otherwise, you'd end up with every tab showing a login page.